hash_practice/exercises.py

```diff
@@ -1,29 +1,99 @@
+from collections import defaultdict
 
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(N)
+        Space Complexity: O(N)
     """
-    pass
+    if len(strings) == 0:
+        return []
+
+    anagrams = {}
+    for s in strings:
+        a = tuple(sorted(s))
+        if a in anagrams:
+            anagrams[a].append(s)
+        else:
+            anagrams[a] = [s]
+    return list(anagrams.values())
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(N)
+        Space Complexity: O(N)
     """
-    pass
+    if len(nums) == 0:
+        return []
 
+    count = {}
+    freq = [[] for i in range(len(nums) + 1)]
+    
+    for n in nums:
+        count[n] = 1 + count.get(n, 0)
+    
+    for n, c in count.items():
+        freq[c].append(n)
+    
+    res = []
+    for i in range(len(freq) - 1, 0, -1):
+        for n in freq[i]:
+            res.append(n)
+            if len(res) == k:
+                return res
 
-def valid_sudoku(table):
+def valid_sudoku(board):
     """ This method will return the true if the table is still
         a valid sudoku table.
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: ON^2
+        Space Complexity: ON+K

```

Comment:
I think this is a perfectly acceptable answer, but I also think it would be valid in this case to give O(1) since the size of our boards are constant.